flink-libraries/flink-table/src/main/java/org/apache/flink/table/runtime/batch/io/RFCCsvInputFormat.java

```diff
@@ -0,0 +1,449 @@
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.StringUtils;
+
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvParser;
+import com.fasterxml.jackson.dataformat.csv.CsvSchema;
+
+import org.apache.commons.io.input.BoundedInputStream;
+
+import java.io.IOException;
+import java.sql.Date;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * InputFormat that reads csv files and compliant with RFC 4180 standards.
+ *
+ * @param <OUT>
+ */
+@Internal
+public abstract class RFCCsvInputFormat<OUT> extends FileInputFormat<OUT> {
+
+	private static final long serialVersionUID = 1L;
+
+	public static final String DEFAULT_LINE_DELIMITER = "\n";
+
+	public static final String DEFAULT_FIELD_DELIMITER = ",";
+
+	private String recordDelimiter = "\n";
+	private char fieldDelimiter = ',';
+
+	private boolean skipFirstLineAsHeader;
+	private boolean skipFirstLine = false; // only for first split
+
+	private boolean quotedStringParsing = false;
+	private char quoteCharacter;
+
+	private boolean lenient;
+
+	private String commentPrefix = null;
+	private boolean allowComments = false;
+
+	private static final Class<?>[] EMPTY_TYPES = new Class<?>[0];
+
+	private static final boolean[] EMPTY_INCLUDED = new boolean[0];
+
+	private Class<?>[] fieldTypes = EMPTY_TYPES;
+
+	private boolean[] fieldIncluded = EMPTY_INCLUDED;
+
+	MappingIterator<Object[]> recordIterator = null;
+
+	private boolean endOfSplit = false;
+
+	protected RFCCsvInputFormat(Path filePath) {
+		super(filePath);
+	}
+
+	@Override
+	public void open(FileInputSplit split) throws IOException {
+
+		super.open(split);
+
+		CsvMapper mapper = new CsvMapper();
+		mapper.disable(CsvParser.Feature.WRAP_AS_ARRAY);
+
+		long firstDelimiterPosition = findFirstDelimiterPosition();
+		long lastDelimiterPosition = findLastDelimiterPosition();
+		long startPos = this.splitStart + firstDelimiterPosition;
+		long endPos = this.splitLength + lastDelimiterPosition - firstDelimiterPosition;
+		this.stream.seek(startPos);
+		BoundedInputStream boundedInputStream = new BoundedInputStream(this.stream, endPos);
+
+		if (skipFirstLineAsHeader && startPos == 0) {
+			skipFirstLine = true;
+		}
+
+		CsvParser csvParser = mapper.getFactory().createParser(boundedInputStream);
+		CsvSchema csvSchema = configureParserSettings();
+		csvParser.setSchema(csvSchema);
+
+		recordIterator = mapper.readerFor(Object[].class).readValues(csvParser);
+	}
+
+	private CsvSchema configureParserSettings() {
+
+		CsvSchema csvSchema = CsvSchema.builder()
+			.setLineSeparator(this.recordDelimiter)
+			.setColumnSeparator(this.fieldDelimiter)
+			.setSkipFirstDataRow(skipFirstLine)
+			.setQuoteChar(this.quoteCharacter)
+			.setAllowComments(allowComments)
+			.build();
```

Comment:
Using `Builder.setColumnType()` we can only specify NUMBER, STRING and BOOLEAN. So, manual casting is required for other data types. Also in this way, we have to specify all columns in schema, and parser doesn't automatically skip fields which are not configured and return a map iterator as a result. 
Also, I think conversion from Array(current implementation) to Row is more efficient rather than from Map to Row. What do you think?